src/libpgmoneta/utils.c

```diff
@@ -3003,6 +3004,198 @@ pgmoneta_is_file_archive(char* file_path)
    return false;
 }
 
+int
+pgmoneta_token_bucket_init(struct token_bucket* tb, unsigned long burst, unsigned long max_rate, int every)
+{
+   if (tb != NULL)
+   {
+      tb->burst = burst;
+      atomic_init(&tb->cur_tokens, burst);
+      tb->max_rate = max_rate;
+      tb->every = every;
+      atomic_init(&tb->last_time, (unsigned long)time(NULL));
+      return 0;
+   }
+   return 1;
+}
+
+void
+pgmoneta_token_bucket_destroy(struct token_bucket* tb)
+{
+   if (tb)
+   {
+      free(tb);
+   }
+}
+
+int
+pgmoneta_token_bucket_add(struct token_bucket* tb)
+{
+   unsigned long diff;
+   unsigned long expected_tokens;
+   unsigned long new_tokens;
+   unsigned long expected_time;
+   unsigned long cur_time;
+
+   diff = (unsigned long) time(NULL) - atomic_load(&tb->last_time);
+
+   if (diff <= 0)
+   {
+      return 0;
+   }
+
+   expected_tokens = atomic_load(&tb->cur_tokens);
+   new_tokens = expected_tokens + tb->max_rate * diff / tb->every;
```

Comment:
I fear it's probably necessary. If diff is smaller than every, then no tokens are added if tokens is in integer. But you still update last_time.., which means if you wait for time shorter than every, you will wait forever, since diff is always smaller than every.

Reply:
Yes, that approach is better for controlling the limit, similar to the nanosecond version. However, I think it would be too frequent to manipulate the token. For us, a rough limit should suffice. This way, we can achieve an average limit with less manipulation, which I believe is more efficient.

Reply:
No, if it is smaller, I will just return 0, quit this addition, and not update the time.

Reply:
Sounds good. Let's stick to your way then.